File: 70-climbing-stairs/70-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        HashMap<Integer,Integer> hm = new HashMap<>();
        return climb(n,hm);
    }
    
    public int climb(int n,HashMap<Integer,Integer> hm) {
        if(hm.containsKey(n)) return hm.get(n);
        if(n==0||n==1) return 1;
        int result = climb(n-1,hm)+climb(n-2,hm);
        hm.put(n,result);
        return result;
    }
}

File: 70-climbing-stairs/70-climbing-stairs.js
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function(n) {
    const dp = [];
    return helper(n, dp);
};

function helper(n, dp){
    if(n===0) return 1;
    if(n<0) return 0;
    if (dp[n] > 0) return dp[n];
    let result = helper(n-1, dp) + helper(n-2, dp);
        return dp[n]=result;
}